Implement v5 transaction consensus rules

LarryRuane opened the pull request:
Closes zcash/zcash#5196.

nuttycom commented on src/gtest/test_checktransaction.cpp:
```diff
@@ -60,6 +60,10 @@ CMutableTransaction GetValidTransaction(uint32_t consensusBranchId=SPROUT_BRANCH
         mtx.fOverwintered = true;
         mtx.nVersionGroupId = SAPLING_VERSION_GROUP_ID;
         mtx.nVersion = SAPLING_TX_VERSION;
+    } else if (consensusBranchId == NetworkUpgradeInfo[Consensus::UPGRADE_NU5].nBranchId) {
+        mtx.fOverwintered = true;
+        mtx.nVersionGroupId = ZIP225_VERSION_GROUP_ID;
+        mtx.nVersion = ZIP225_TX_VERSION;
```
Here in the tests it doesn't really matter, right?

str4d replied:
Oh, I missed that this was the tests. There is very similar code used by the miner for coinbase transactions.

str4d commented on src/primitives/transaction.h:
```diff
@@ -374,6 +374,14 @@ class OrchardBundle
         inner.reset(bundle);
     }
+
+    bool outputs_enabled() {
+        return orchard_bundle_outputs_enabled(inner.get());
+    }
+
+    bool spends_enabled() {
+        return orchard_bundle_spends_enabled(inner.get());
+    }
```
This PR will need to be rebased on master, as `OrchardBundle` moved to `src/primitives/orchard.h`.

LarryRuane commented:
Rebased to resolve merge conflicts. Also added calls in test code to the new methods, `outputs_enabled()` and `spends_enabled()` to make sure the linkages work. These will be tested more thoroughly later.

LarryRuane commented:
Rebased to pick up #5227 changes, helpful to this PR because of the addition of `tx.GetOrchardBundle().IsPresent()` among other methods.

nuttycom commented on src/main.cpp:
```diff
@@ -1080,7 +1134,9 @@ bool ContextualCheckTransaction(
         !tx.vShieldedSpend.empty() ||
-        !tx.vShieldedOutput.empty())
+        !tx.vShieldedOutput.empty() ||
+        orchard_bundle.SpendsEnabled() ||
```
We know by construction that if an orchard bundle is present that it contains a nonempty list of actions, so it's okay to just check the flags here, but I'd appreciate a comment to this effect; I had forgotten whether that was the case and had to go look it up.

nuttycom commented on src/main.cpp:
```diff
@@ -1280,17 +1340,28 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
+    if (tx.vin.empty() &&
+        tx.vJoinSplit.empty() &&
+        tx.vShieldedSpend.empty() &&
+        !orchard_bundle.SpendsEnabled())
```
I guess these checks are really just defense-in-depth, because they don't really guarantee anything other than a structural validity property. IMO that's more a part of parsing than an important consensus rule; zero-valued notes are possible, right? So this is just a fail-fast filter before we go to check the value balance?

str4d replied:
Correct, this is a fail-fast check that rejects obviously-invalid transaction constructions that the parser doesn't structurally prevent.

daira replied:
This doesn't seem to implement the specified consensus rule, because it doesn't check `nActionsOrchard`.

nuttycom reviewed (changes requested):
We need to make sure that Sapling V4 transactions are still accepted as valid after NU5; also, this needs tests that validate against Orchard data.

nuttycom commented on src/main.cpp:
```diff
@@ -894,23 +903,15 @@ bool ContextualCheckTransaction(
             }
         }
-
-        if (!futureActive) {
```
This check is still valid during the pre-NU5 epoch.

daira commented on src/coins.h:
```diff
@@ -313,10 +313,14 @@ struct CNullifiersCacheEntry
 
+/// These identify the value pool, and as such, Canopy (for example)
+/// isn't here, since value created during the Canopy network upgrade
+/// is part of the Saplingn pool.
```
```suggestion
/// is part of the Sapling pool.
```

nuttycom reviewed (commented):
Flushing comments.

str4d commented on src/main.cpp:
```diff
@@ -1045,19 +1052,125 @@ bool ContextualCheckTransaction(
+        // Rules that apply generally before Nu5. These were
+        // previously noncontextual checks that became contextual
+        // after Nu5 activation.
+        if (orchard_bundle.IsPresent()) {
+            return state.DoS(
+                dosLevelPotentiallyRelaxing,
+                error("ContextualCheckTransaction(): transaction has Orchard actions"),
+                REJECT_INVALID, "bad-tx-has-orchard-actions");
+        }
```
I'd argue that this is also an internal `zcashd` consistency check (since similarly to `nConsensusBranchId`, this can't occur for pre-v5 transactions from the network).

AFAICT we don't have an equivalent consistency check for Sapling, but I'm happy to have these here.

str4d commented on src/main.cpp:
```diff
@@ -1045,19 +1052,125 @@ bool ContextualCheckTransaction(
+            // tx.nConsensusBranchId must match the current consensus branch id
+            if (!(tx.GetConsensusBranchId() && *tx.GetConsensusBranchId() == consensusBranchId)) {
+                return state.DoS(
+                    dosLevelPotentiallyRelaxing,
```
I believe this makes sense here. If we see a mempool transaction when the chain tip is post-NU6 but we are in initial chain sync at the NU5 epoch, the transaction would fail this check, but would pass it at a later height.

str4d commented on src/main.cpp:
```diff
@@ -1077,14 +1193,16 @@ bool ContextualCheckTransaction(
         // Rules that apply generally before the next release epoch
     }
 
-    auto consensusBranchId = CurrentEpochBranchId(nHeight, consensus);
     auto prevConsensusBranchId = PrevEpochBranchId(consensusBranchId, consensus);
     uint256 dataToBeSigned;
     uint256 prevDataToBeSigned;
 
+    // Create signature hashes for shielded components.
     if (!tx.vJoinSplit.empty() ||
         !tx.vShieldedSpend.empty() ||
-        !tx.vShieldedOutput.empty())
+        !tx.vShieldedOutput.empty() ||
+        orchard_bundle.SpendsEnabled() ||
+        orchard_bundle.OutputsEnabled())
```
Revert this change; we don't check Orchard signatures here, and it just causes us to unnecessarily call `SignatureHash` twice on pure-Orchard v5 transactions.

str4d commented on src/main.cpp:
```diff
@@ -1284,17 +1404,28 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
     }
+    auto orchard_bundle = tx.GetOrchardBundle();
 
     // Transactions containing empty `vin` must have either non-empty
     // `vJoinSplit` or non-empty `vShieldedSpend`.
```
```suggestion
    // Transactions must contain some potential source of funds. This
    // rejects obviously-invalid transaction constructions early, but
    // cannot prevent e.g. a pure Sapling transaction with only dummy
    // spends (which is undetectable).
```

str4d commented on src/main.cpp:
```diff
@@ -1284,17 +1404,28 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
         return state.DoS(10, error("CheckTransaction(): vin empty"),
                          REJECT_INVALID, "bad-txns-vin-empty");
+    }
     // Transactions containing empty `vout` must have either non-empty
     // `vJoinSplit` or non-empty `vShieldedOutput`.
```
```suggestion
    // Transactions must contain some potential useful sink of funds.
    // This rejects obviously-invalid transaction constructions early,
    // but cannot prevent e.g. a pure Sapling transaction with only
    // dummy outputs (which is undetectable), and does not prevent
    // transparent transactions from sending all funds to miners.
```

str4d commented on src/main.cpp:
```diff
@@ -1319,28 +1450,50 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
         }
     }
 
+    // Check for non-zero valueBalanceOrchard when there are no Orchard inputs or outputs
+    if (!orchard_bundle.SpendsEnabled() && !orchard_bundle.OutputsEnabled() && orchard_bundle.GetValueBalance() != 0) {
```
Cache this to avoid crossing the FFI every time:
```suggestion
    auto valueBalanceOrchard = orchard_bundle.GetValueBalance();

    // Check for non-zero valueBalanceOrchard when there are no Orchard inputs or outputs
    if (!orchard_bundle.SpendsEnabled() && !orchard_bundle.OutputsEnabled() && valueBalanceOrchard != 0) {
```
(This kind of refactor would hopefully become unnecessary once we can use cross-language LTO successfully.)

str4d commented on src/main.cpp:
```diff
@@ -1319,28 +1450,50 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
+    }
+
+    // Check for overflow valueBalanceOrchard
+    if (orchard_bundle.GetValueBalance() > MAX_MONEY || orchard_bundle.GetValueBalance() < -MAX_MONEY) {
```
```suggestion
    if (valueBalanceOrchard > MAX_MONEY || valueBalanceOrchard < -MAX_MONEY) {
```

str4d commented on src/main.cpp:
```diff
@@ -1319,28 +1450,50 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
+                         REJECT_INVALID, "bad-txns-valuebalance-toolarge");
+    }
+
+    if (orchard_bundle.GetValueBalance() <= 0) {
+        // NB: negative valueBalanceOrchard "takes" money from the transparent value pool just as outputs do
+        nValueOut += -orchard_bundle.GetValueBalance();
```
```suggestion
    if (valueBalanceOrchard <= 0) {
        // NB: negative valueBalanceOrchard "takes" money from the transparent value pool just as outputs do
        nValueOut += -valueBalanceOrchard;
```

str4d commented on src/main.cpp:
```diff
@@ -1405,6 +1569,9 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
     }
 
     // Check for duplicate inputs
+    //
+    // A transaction with one or more transparent inputs from coinbase transactions
+    // MUST have no transparent outputs (i.e. tx_out_count MUST be 0)
```
This is not where this check happens (we only know the types of the inputs in `ContextualCheckInputs`, which is by definition contextual).
```suggestion
```

str4d commented on src/main.cpp:
```diff
@@ -1045,19 +1052,125 @@ bool ContextualCheckTransaction(
+        }
+
+        if (tx.IsCoinBase()) {
+            // A coinbase transaction cannot have Orchard spends.
+            if (orchard_bundle.SpendsEnabled()) {
+                return state.DoS(
+                    dosLevelPotentiallyRelaxing,
+                    error("ContextualCheckTransaction(): coinbase has orchard shielded spends"),
+                    REJECT_INVALID, "bad-cb-has-orchard-spends");
+            }
+
```
This is being checked non-contextually (in `CheckTransactionWithoutProofVerification` below).
```suggestion
```

str4d commented on src/main.cpp:
```diff
@@ -4112,13 +4282,13 @@ bool ReceivedBlockTransactions(
+        saplingValue += -tx.GetValueBalanceSapling();
 
-        // Orchard valueBalance behaves the same way as Sapling valueBalance.
+        // Orchard valueBalanceSapling behaves the same way as Sapling valueBalanceSapling.
```
Regex mismatch, but we can clean this up further:
```suggestion
        // valueBalanceOrchard behaves the same way as valueBalanceSapling.
```

daira commented on src/main.cpp:
```diff
@@ -1045,19 +1053,123 @@ bool ContextualCheckTransaction(
+            // transactions.  NOTE: This is an internal zcashd consistency
+            // check; it does not correspond to a consensus rule in the
+            // protocol specification, but is instead an artifact of the
+            // internal zcashd transaction representation.
```
```suggestion
            // NOTE: This is an internal zcashd consistency check;
            // it does not correspond to a consensus rule in the
            // protocol specification, but is instead an artifact of
            // the internal zcashd transaction representation.
```

daira commented on src/main.cpp:
```diff
@@ -1045,19 +1053,123 @@ bool ContextualCheckTransaction(
+            if (!(tx.GetConsensusBranchId() && *tx.GetConsensusBranchId() == consensusBranchId)) {
+                return state.DoS(
+                    dosLevelPotentiallyRelaxing,
+                    error("ContextualCheckTransaction(): consensus branch id not set"),
+                    REJECT_INVALID, "bad-tx-missing-consensus-branch-id");
```
"consensus branch id not set" and "bad-tx-missing-consensus-branch-id" both suggest that this error only covers a missing branch id, rather than an incorrect branch id.

daira commented on src/main.cpp:
```diff
@@ -1045,19 +1053,123 @@ bool ContextualCheckTransaction(
+            }
+        }
+
+        //nSpendsSapling, nOutputsSapling, and nActionsOrchard MUST all be less than 2^16
```
```suggestion
        // nSpendsSapling, nOutputsSapling, and nActionsOrchard MUST all be less than 2^16
```

daira commented on src/main.cpp:
```diff
@@ -1045,19 +1053,123 @@ bool ContextualCheckTransaction(
+        }
+
+        //nSpendsSapling, nOutputsSapling, and nActionsOrchard MUST all be less than 2^16
+        size_t max_spends = 1 << 16;
```
This variable is misnamed, for two reasons:
* it's used for outputs and actions as well as spends;
* it's not the maximum, it's the maximum plus one.

daira commented on src/main.cpp:
```diff
@@ -1045,19 +1053,123 @@ bool ContextualCheckTransaction(
+        if (tx.vShieldedSpend.size() >= max_spends) {
+            return state.DoS(
+                dosLevelPotentiallyRelaxing,
+                error("ContextualCheckTransaction(): More than 2^16 Sapling spends"),
```
```suggestion
                error("ContextualCheckTransaction(): 2^16 or more Sapling spends"),
```

daira commented on src/main.cpp:
```diff
@@ -1045,19 +1053,123 @@ bool ContextualCheckTransaction(
+        if (tx.vShieldedOutput.size() >= max_spends) {
+            return state.DoS(
+                dosLevelPotentiallyRelaxing,
+                error("ContextualCheckTransaction(): More than 2^16 Sapling outputs"),
```
```suggestion
                error("ContextualCheckTransaction(): 2^16 or more Sapling outputs"),
```

daira commented on src/main.cpp:
```diff
@@ -1045,19 +1053,123 @@ bool ContextualCheckTransaction(
+        if (orchard_bundle.GetNumActions() >= max_spends) {
+            return state.DoS(
+                dosLevelPotentiallyRelaxing,
+                error("ContextualCheckTransaction(): More than 2^16 Orchard actions"),
```
```suggestion
                error("ContextualCheckTransaction(): 2^16 or more Orchard actions"),
```

daira commented on src/main.cpp:
```diff
@@ -1045,19 +1053,123 @@ bool ContextualCheckTransaction(
+            // Check that Orchard coinbase outputs can be decrypted with the all-zeros OVK
+        }
+    } else {
+        // Rules that apply generally before Nu5. These were
+        // previously noncontextual checks that became contextual
+        // after Nu5 activation.
```
```suggestion
        // Rules that apply generally before NU5. These were previously
        // noncontextual checks that became contextual after NU5 activation.
```

daira commented on src/main.cpp:
```diff
@@ -1045,19 +1053,123 @@ bool ContextualCheckTransaction(
+        }
+
+        if (tx.IsCoinBase()) {
+            // Check that Orchard coinbase outputs can be decrypted with the all-zeros OVK
```
```suggestion
            // TODO: Check that Orchard coinbase outputs can be decrypted with the all-zeros OVK
```

daira commented on src/main.cpp:
```diff
@@ -1284,17 +1403,33 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
+    if (tx.vout.empty() &&
+        tx.vJoinSplit.empty() &&
+        tx.vShieldedOutput.empty() &&
+        !orchard_bundle.OutputsEnabled())
```
This doesn't seem to implement the specified consensus rule, because it doesn't check `nActionsOrchard`.

nuttycom replied:
I should add a comment - these calls that look like flag checks are confusing, because they also check that the Orchard bundle is present. If an Orchard bundle exists, its actions vector is nonempty by construction.

str4d commented on src/main.cpp:
```diff
@@ -1102,23 +1102,32 @@ bool ContextualCheckTransaction(
                     REJECT_INVALID, "bad-tx-missing-consensus-branch-id");
             }
+
+            // v5 transactions must have empty joinSplits
```
I think it's fine to put this consistency check here. If NU5 is not active, then v5 transactions will be unilaterally rejected, and if NU5 is active, then we reach this check.

str4d reviewed (approved):
utACK. After this PR, I believe we are missing two consensus rules:
- ZIP 213 changes (#5243).
- ZIP 203 changes (I'm making a PR for these).

daira commented on src/main.cpp:
```diff
@@ -1319,28 +1454,52 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
+    auto valueBalanceOrchard = orchard_bundle.GetValueBalance();
+
+    // Check for non-zero valueBalanceOrchard when there are no Orchard inputs or outputs
+    if (!orchard_bundle.SpendsEnabled() && !orchard_bundle.OutputsEnabled() && valueBalanceOrchard != 0) {
```
There is currently no such rule in the spec, although I agree it makes sense.

daira replied:
https://github.com/zcash/zips/issues/531

daira commented on src/main.cpp:
```diff
@@ -1468,6 +1638,9 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
                              REJECT_INVALID, "bad-cb-has-spend-description");
         // See ContextualCheckTransaction for consensus rules on coinbase output descriptions.
+        if (orchard_bundle.SpendsEnabled())
+            return state.DoS(100, error("CheckTransaction(): coinbase has Orchard spends"),
```
```suggestion
            return state.DoS(100, error("CheckTransaction(): coinbase has enableSpendsOrchard set"),
```

daira reviewed (changes requested):
Reviewed.

nuttycom commented on src/gtest/test_checktransaction.cpp:
```diff
@@ -58,6 +58,7 @@ class MockCValidationState : public CValidationState {
 
 void CreateJoinSplitSignature(CMutableTransaction& mtx, uint32_t consensusBranchId);
 
+// TODO If creating a zip225 (v5) tx, we need to be told whether to create sprout elements or orchard elements.
```
Oh, I'd stripped out a couple of the other changes to this file but forgot to strip out the comment.

nuttycom commented on src/main.cpp:
```diff
@@ -1045,19 +1053,132 @@ bool ContextualCheckTransaction(
+            if (!(tx.GetConsensusBranchId() && *tx.GetConsensusBranchId() == consensusBranchId)) {
+                return state.DoS(
+                    dosLevelPotentiallyRelaxing,
+                    error("ContextualCheckTransaction(): consensus branch id not set"),
+                    REJECT_INVALID, "bad-tx-missing-consensus-branch-id");
```
```suggestion
                    error("ContextualCheckTransaction(): transaction's consensus branch id does not match the current consensus branch"),
                    REJECT_INVALID, "bad-tx-consensus-branch-id-mismatch");
```

daira commented on src/main.cpp:
```diff
@@ -1284,17 +1403,33 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
+    // rejects obviously-invalid transaction constructions early, but
+    // cannot prevent e.g. a pure Sapling transaction with only dummy
+    // spends (which is undetectable).
+    if (tx.vin.empty() &&
+        tx.vJoinSplit.empty() &&
+        tx.vShieldedSpend.empty() &&
+        !orchard_bundle.SpendsEnabled())
```
We should closely match the spec. It greatly simplifies checking consistency between them.

daira commented on src/main.cpp:
```diff
@@ -1045,19 +1053,131 @@ bool ContextualCheckTransaction(
+
+
+        // nSpendsSapling, nOutputsSapling, and nActionsOrchard MUST all be less than 2^16
+        size_t max_spends = (1 << 16) - 1;
+        if (tx.vShieldedSpend.size() > max_spends) {
+            return state.DoS(
+                dosLevelPotentiallyRelaxing,
+                error("ContextualCheckTransaction(): 2^16 or more Sapling spends"),
+                REJECT_INVALID, "bad-tx-too-many-sapling-spends");
+        }
+        if (tx.vShieldedOutput.size() > max_spends) {
+            return state.DoS(
+                dosLevelPotentiallyRelaxing,
+                error("ContextualCheckTransaction(): 2^16 or more Sapling outputs"),
+                REJECT_INVALID, "bad-tx-too-many-sapling-outputs");
+        }
+        if (orchard_bundle.GetNumActions() > max_spends) {
```
```suggestion
        size_t max_elements = (1 << 16) - 1;
        if (tx.vShieldedSpend.size() > max_elements) {
            return state.DoS(
                dosLevelPotentiallyRelaxing,
                error("ContextualCheckTransaction(): 2^16 or more Sapling spends"),
                REJECT_INVALID, "bad-tx-too-many-sapling-spends");
        }
        if (tx.vShieldedOutput.size() > max_elements) {
            return state.DoS(
                dosLevelPotentiallyRelaxing,
                error("ContextualCheckTransaction(): 2^16 or more Sapling outputs"),
                REJECT_INVALID, "bad-tx-too-many-sapling-outputs");
        }
        if (orchard_bundle.GetNumActions() > max_elements) {
```

daira commented on src/main.cpp:
```diff
@@ -1284,17 +1411,35 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
+    // prevent e.g. a pure Sapling transaction with only dummy outputs (which
+    // is undetectable), and does not prevent transparent transactions from
+    // sending all funds to miners.  Contextual checks ensure that only one of
+    // Sprout joinsplits or Orchard actions may be present.
```
```suggestion
    // Sprout joinsplits or Orchard actions may be present.
    // Note that orchard_bundle.OutputsEnabled() is false when
    // there is no Orchard bundle, i.e. when nActionsOrchard == 0.
```

daira reviewed (commented):
The nExpiryHeight changes made in v2021.2.6 (and in ZIP 203) don't seem to have been included:

![image](https://user-images.githubusercontent.com/643204/124032714-d55dba00-d9f0-11eb-9836-feda15d026c8.png)

str4d commented:
@daira I noted that in my review; I have a PR ready to go once this PR merges, as that's more of a block rule, not a transaction rule (the change goes into `ContextualCheckBlock`).

str4d commented:
Actually, the ZIP 203 change is small, and since CI takes an hour, I decided to push the commit to this PR.

daira commented on src/main.cpp:
```diff
@@ -829,7 +831,10 @@ bool ContextualCheckTransaction(
     assert(!heartwoodActive || saplingActive);  // Heartwood cannot be active unless Sapling is
     assert(!canopyActive || heartwoodActive);   // Canopy cannot be active unless Heartwood is
-    assert(!futureActive || canopyActive);      // ZFUTURE must always include the latest live version
+    assert(!nu5Active || canopyActive);         // Nu5 cannot be active unless Canopy is
```
```suggestion
    assert(!nu5Active || canopyActive);         // NU5 cannot be active unless Canopy is
```

daira commented on src/main.cpp:
```diff
@@ -829,7 +831,10 @@ bool ContextualCheckTransaction(
     assert(!canopyActive || heartwoodActive);   // Canopy cannot be active unless Heartwood is
-    assert(!futureActive || canopyActive);      // ZFUTURE must always include the latest live version
+    assert(!nu5Active || canopyActive);         // Nu5 cannot be active unless Canopy is
+    assert(!futureActive || nu5Active);         // ZFUTURE must always include the latest live version
```
The comment isn't quite accurate. It enforces that ZFUTURE includes NU5, but NU5 isn't live yet.

daira commented on src/main.cpp:
```diff
@@ -1045,19 +1053,143 @@ bool ContextualCheckTransaction(
+        // noncontextual checks that became contextual after NU5 activation.
+
+        if (tx.nExpiryHeight >= TX_EXPIRY_HEIGHT_THRESHOLD) {
+            return state.DoS(100, error("CheckTransaction(): expiry height is too high"),
```
This should be dosLevelPotentiallyRelaxing.

daira commented on src/main.cpp:
```diff
@@ -1275,26 +1413,43 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
+    // obviously-invalid transaction constructions early, but cannot prevent
+    // e.g. a pure Sapling transaction with only dummy spends (which is
+    // undetectable). Contextual checks ensure that only one of Sprout
+    // joinsplits or Orchard actions may be present.
```
```suggestion
    // joinsplits or Orchard actions may be present.
    // Note that orchard_bundle.SpendsEnabled() is false when no
    // Orchard bundle is present, i.e. when nActionsOrchard == 0.
```

daira commented on src/main.cpp:
```diff
@@ -1045,19 +1053,143 @@ bool ContextualCheckTransaction(
+        // noncontextual checks that became contextual after NU5 activation.
+
+        if (tx.nExpiryHeight >= TX_EXPIRY_HEIGHT_THRESHOLD) {
+            return state.DoS(100, error("CheckTransaction(): expiry height is too high"),
+                            REJECT_INVALID, "bad-tx-expiry-height-too-high");
```
```suggestion
            return state.DoS(
                dosLevelPotentiallyRelaxing,
                error("CheckTransaction(): expiry height is too high"),
                REJECT_INVALID, "bad-tx-expiry-height-too-high");
```

daira commented on src/main.cpp:
```diff
@@ -1045,19 +1053,143 @@ bool ContextualCheckTransaction(
+            // transactions.
+            if (tx.nExpiryHeight >= TX_EXPIRY_HEIGHT_THRESHOLD) {
+                return state.DoS(100, error("CheckTransaction(): expiry height is too high"),
+                                REJECT_INVALID, "bad-tx-expiry-height-too-high");
```
```suggestion
                                 REJECT_INVALID, "bad-tx-expiry-height-too-high");
```

daira commented on src/primitives/orchard.h:
```diff
@@ -108,6 +113,18 @@ class OrchardBundle
+        return orchard_bundle_spends_enabled(inner.get());
+    }
+
+    bool HasValidCoinbaseOutputs() const {
```
This returns false if there is no Orchard bundle, but shouldn't it vacuously return true in that case because all Orchard coinbase outputs are valid? In any case, ensure that `orchard_bundle_has_valid_coinbase_outputs` has the desired behaviour for the no-bundle case and drop the `inner &&`.

nuttycom replied:
Ah, good catch.

daira commented on src/rust/include/rust/orchard.h:
```diff
@@ -108,6 +108,19 @@ void orchard_batch_add_bundle(
+bool orchard_bundle_spends_enabled(const OrchardBundlePtr* bundle);
+
+/// Returns whether all actions contained in the Orchard bundle
+/// can be decrypted with the all-zeros OVK.
```
Document the behaviour for the no-bundle case.

daira commented on src/rust/src/orchard_ffi.rs:
```diff
@@ -271,3 +273,47 @@ pub extern "C" fn orchard_batch_validate(batch: *const BatchValidator) -> bool {
+
+    // Either there are no Orchard actions, or all of the outputs
+    // are decryptable with the all-zeros OVK.
+    true
```
I agree that true is the right output for the no-bundle case. Currently `HasValidCoinbaseOutputs()` returns false for that case, though.

daira commented on src/primitives/orchard.h:
```diff
@@ -108,6 +113,18 @@ class OrchardBundle
+        return orchard_bundle_spends_enabled(inner.get());
+    }
+
+    bool HasValidCoinbaseOutputs() const {
```
Consider renaming this to `CoinbaseOutputsAreValid` (to avoid the implication that there must be any such outputs).

daira commented on src/rust/include/rust/orchard.h:
```diff
@@ -108,6 +108,20 @@ void orchard_batch_add_bundle(
+/// Returns whether all actions contained in the Orchard bundle
+/// can be decrypted with the all-zeros OVK. Returns `true`
+/// if no Orchard actions are present.
+bool orchard_bundle_has_valid_coinbase_outputs(const OrchardBundlePtr* bundle);
```
Consider renaming this to `orchard_bundle_coinbase_outputs_are_valid` (to avoid the implication that there must be any such outputs).

nuttycom commented on src/main.cpp:
```diff
@@ -1275,26 +1420,45 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
             return state.DoS(100, error("CheckTransaction(): unknown tx version group id"),
                     REJECT_INVALID, "bad-tx-version-group-id");
         }
-        if (tx.nExpiryHeight >= TX_EXPIRY_HEIGHT_THRESHOLD) {
```
Verified that this was moved to contextual checks.

nuttycom reviewed (approved):
ACK of the parts not implemented by me: https://github.com/zcash/zcash/pull/5225/commits/8ce8db44222d15f0b962093ad63ff0d1cb9985c5

daira commented on src/main.cpp:
```diff
@@ -1275,26 +1420,45 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
+        tx.vShieldedSpend.empty() &&
+        !orchard_bundle.SpendsEnabled())
+    {
         return state.DoS(10, error("CheckTransaction(): vin empty"),
                          REJECT_INVALID, "bad-txns-vin-empty");
```
```suggestion
        return state.DoS(10, error("CheckTransaction(): no source of funds"),
                         REJECT_INVALID, "bad-txns-no-source-of-funds");
```

daira commented on src/main.cpp:
```diff
@@ -1275,26 +1420,45 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
+        tx.vShieldedOutput.empty() &&
+        !orchard_bundle.OutputsEnabled())
+    {
         return state.DoS(10, error("CheckTransaction(): vout empty"),
                          REJECT_INVALID, "bad-txns-vout-empty");
```
```suggestion
        return state.DoS(10, error("CheckTransaction(): no sink of funds"),
                         REJECT_INVALID, "bad-txns-no-sink-of-funds");
```

daira commented on src/main.cpp:
```diff
@@ -1393,9 +1581,20 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
+
+            if (!MoneyRange(nValueIn)) {
+                return state.DoS(100, error("CheckTransaction(): txin total out of range"),
+                                    REJECT_INVALID, "bad-txns-txintotal-toolarge");
```
```suggestion
                                 REJECT_INVALID, "bad-txns-txintotal-toolarge");
```

daira reviewed (approved):
utACK 🎉

str4d reviewed (approved):
utACK

str4d commented:
@zkbot r+

daira reviewed (approved):
Also utACK for the coinbase `nExpiryHeight` check.

nuttycom commented:
@zkbot try

str4d commented:
Force-pushed to clean up the commit history. [The force-push is a no-op](https://github.com/zcash/zcash/compare/0cb4c5cc2c2e29b29b2f71acda4d924021e0af3e..7a2be8876886c0ecb400e50b448b0e5b37c9b014).

str4d reviewed (approved):
Tested ACK.

str4d commented:
@zkbot r+

nuttycom reviewed (approved):
utACK e18da344

Reviewed the last 3 commits.
